### ISSUE TITLE
A new way to expose metrics 

### ISSUE THREAD

**Issue by ZENOTME** [icelake-io/icelake#253]
## Problem
For our writer, we hope to expose our metrics but not bind to the specific lib such as Prometheus. 

## Solution
Here is the trait we can used it to let user custom their updater. The updater will receive the metrics with id which indicates that it belong to which writer. And user can custom the updater logic such as using the Prometheus or just print it as log.
```
pub trait MetricsUpdater<M: Metrics> {
   fn update(&mut self, metrics: &M);
}

pub trait Metrics {
   fn id(&self) -> size;
}

// Upadter will be used like following:
impl RollingWriterBuilder<U: MetricsUpdater<M>> {
   pub fn new(..,updater: U) 
}

struct RollingWriter<U: MetricsUpdater<M>> {
  metrics: RollingWriterMetrics,
  updater: U
}

impl RollingWriter {
  pub fn write(&mut self) {
     ...
     // update metrics
    self.updater.update(self.metrics)
  }
}
```
![image](https://github.com/icelake-io/icelake/assets/43447882/a0cb59b8-fb63-46c5-9dac-54a1418c4821)

## Drawback 
Every writer have to bind with a id because we need identify metrics from different writer. Seems there is no other way to avoid this in this solution. 

cc @liurenjie1024  @Xuanwo Feel free for any suggestions or better design.

**Comment by liurenjie1024** [comment]
After the refactoring, I think maybe we can have another approach:
```rust
struct RollingMetrics {
  unflush_data_file: usize
}

struct RollingWriter {
   fn metrics(&self) -> RollingMetrics {}
}

struct RollingWriterWithMetrics {
  metrics: IntGauge,
  writer: RollingWriter,
}

impl IcebergWriter for RollingWriterWithMetrics 
{
}
```

**Comment by ZENOTME** [comment]
> After the refactoring, I think maybe we can have another approach:
> 
> ```rust
> struct RollingMetrics {
>   unflush_data_file: usize
> }
> 
> struct RollingWriter {
>    fn metrics(&self) -> RollingMetrics {}
> }
> 
> struct RollingWriterWithMetrics {
>   metrics: IntGauge,
>   last_rolling_metrics: Option<RollingMetrics>,
>   writer: RollingWriter,
> }
> 
> impl IcebergWriter for RollingWriterWithMetrics 
> {
> }
> ```
> 
> This way we don't need to let user implement metrics updater, and no id required.

LGTM!

**Comment by ZENOTME** [comment]
https://github.com/icelake-io/icelake/pull/254#discussion_r1416570482

Another way is to separate the metrics into per writer. (This needs a more complex design: the writer ID and register in the builder. So that gives users more flexibility: they can get the per-writer metrics and compute the sum in Prometheus. This is a point that can be improved. 🤔 @liurenjie1024

**Comment by liurenjie1024** [comment]
Yes, that's possible, but more detail metrics means more storage in time. Currently there is not much need for this, and we can add this later when necessary.
